printer.go

```diff
@@ -0,0 +1,300 @@
+// Copyright (C) 2023 by Posit Software, PBC
+package rsf
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+)
+
+func Print(w io.Writer, r *bufio.Reader, o any) error {
+	// Create a new reader since we need to read the RSF data.
+	reader := NewReader()
+
+	// Read the RSF index. We'll use this data to help print the information.
+	idx, err := reader.ReadIndex(r)
+	if err != nil {
+		return fmt.Errorf("error reading index: %s", err)
+	}
+
+	// Also extract information about array indexes from the object. This
+	// information is not available in the index.
+	sizes := make(map[string]int)
+	kinds := make(map[string]reflect.Kind)
+	err = indexArrays("", "", reflect.TypeOf(o), sizes, kinds)
+	if err != nil {
+		return fmt.Errorf("error indexing arrays: %s", err)
+	}
+
+	// Iterate the fields recursively and print the data.
+	var i int
+	for {
+		i++
+
+		// Read full object size
+		_, err = reader.ReadSizeField(r)
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+
+		// Add blank newline unless at first object
+		if i > 1 {
+			_, err = fmt.Fprintln(w, "")
+			if err != nil {
+				return nil
+			}
+		}
+
+		// Print object header
+		pad := strings.Repeat(" ", 16)
+		header := fmt.Sprintf("%s%s[%d]%s", pad, reflect.TypeOf(o).Name(), i, pad)
+		line := strings.Repeat("-", len(header))
+		_, err = fmt.Fprintf(w, "%s\n%s\n%s\n", line, header, line)
+		if err != nil {
+			return nil
+		}
+
+		// Print data for each field of the object.
+		for _, f := range idx {
+			err = printField(sizes, kinds, "", f, w, r, reader, 0)
+			if err != nil {
+				if err == io.EOF {
+					return nil
+				}
+				return fmt.Errorf("error printing data: %s", err)
+			}
+		}
+	}
+}
+
+func indexArrays(parent, parentIndex string, v reflect.Type, sizes map[string]int, kinds map[string]reflect.Kind) error {
+	if v.Kind() == reflect.Struct {
+		for i := 0; i < v.NumField(); i++ {
+			t := &tag{}
+			_, err := getTagInfo(v, i, t, &tag{}, "")
+			if err != nil {
+				return err
+			}
+			key := t.name
+			if parent != "" {
+				key = strings.Join([]string{parent, t.name}, "...")
+			}
+			sub := v.Field(i)
+			if parentIndex == t.name {
+				if t.fixed > 0 {
+					sizes[parent] = t.fixed
+					kinds[parent] = reflect.String
+				}
+				switch sub.Type.Kind() {
+				case reflect.Int, reflect.Int64, reflect.Int32, reflect.Int16, reflect.Int8:
+					sizes[parent] = 10
+					kinds[parent] = reflect.Int64
+				}
```

Comment:
It could be, but it's not really required. If the type is integer, then `fixed` will be zero, or at least it should be. In the odd case that someone included a `fixed` struct tag with an incorrect value, then _not_ having an `else` will ensure that we set the correct size for an integer index rather than using the incorrect `fixed` value.

Reply:
I added a comment to explain this logic, but let me know if you disagree!